Add MLIP Tutorial

PythonFZ opened the pull request:
Add a Tutorial on Machine-Learning Interatomic Potentials.

RudolfWeeber commented:
Thank you, Fabian.
@SamTov agreed to do a detailed review next week.

Whlie having a quick look, I noticed the following:
* there are some terms and abbreviaitons which people new to the field might not know. A bit more contet and explanaiton in the intro would be useful
* Maybe the role of the different external parckges could be stated mor explicilty. In particular, how Espresso, ASe and Mace interact
* Units: It wans't completely clear to me at first glance what units are used in the simulatoion. Is ESPResSo run in the usual reduced units? 
* As long as there are only external forces, the Verlet-Lists and skin can//should eb switche off
* It would be important to get the external depencies stable, as we will have to include them in the CI

PythonFZ commented:
Hey Rudolf,

Thanks for the feedback. I have added some more clarification w.r.t. to the code that is used and added some more sources. I'd appreciate if you could help me improve which parts still need more clarification.

> * Units: It wans't completely clear to me at first glance what units are used in the simulatoion. Is ESPResSo run in the usual reduced units?

I have talked to @reinaual about the units and if I understood it correctly, I am defining the unit system in ESPResSo such that it resembles the one used within ASE.

> * As long as there are only external forces, the Verlet-Lists and skin can//should eb switche off

As I am not an ESPResSo user, would it be possible to give me an example or edit the file direclty.

> * It would be important to get the external depencies stable, as we will have to include them in the CI

I have removed the `ipsuite` dependency because all the features required are within the `rdkit2ase` package.
The new dependencies would be:
- `pip install mace-torch==0.3.6`
- `pip install rdkit2ase==0.1.4`
- https://github.com/m3g/packmol/releases/tag/v20.15.1

SamTov commented on doc/tutorials/mlip/mlip.ipynb:
```diff
@@ -0,0 +1,447 @@
+{
```
<ul><li>I think the title should either be Machine-Learned Interatomic Potentials or Machine Learning for Interatomic Potentials.</li><li>Maybe mention what kind model is? Something like, Equivariant graph model would be sufficient along with a link to the model paper. This is just the foundation model paper but a direct reference to the architecture would be nice too.</li><li>Do you have an overview of the ASE features somewhere?</li></ul><p> </p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4982/discussion/#comment-1730756644'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='0'/>

SamTov replied:
<p>Why not explain each import to highlight what it is and why you need it? You can start with a code block and then write out a list of why you need each one. For example, pandas is not a clear thing to import in a simulation study unless you need it somewhere. I also usually split my imports by category in tutorials, something like:</p><p> </p>

<pre class="ql-syntax" spellcheck="false"># Simulation imports
import espressomd
from espressomd.plugins.ase import ASEInterface
from mace.calculators import mace_mp
import pint

# Linalg
import numpy as np
import pandas as pd

# Plotting and Visualisation
import plotly.express as px
from espressomd.zn import Visualizer

# Utils
import tqdm

# Analysis
from rdkit2ase import smiles2atoms, pack
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4982/discussion/#comment-1730756647'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='2'/>

SamTov replied:
<p>You can use full variable names to alleviate any uncertainty. Things aren't clear to everyone. For me, kb would be clearer than boltmk but as you never introduce the fact that you will talk about Boltzmann, I would recommend just writing out boltzmann_constant or boltzmann_k. Maybe also add comments after each line highlight what unit conversions you are doing.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4982/discussion/#comment-1730756649'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='4'/>

SamTov replied:
<p>This is pretty random after you talk about SMILES strings. Why not have a section about what this initialisation does and then show it. After, introduce SMILES and do the system construction.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4982/discussion/#comment-1730756650'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='6'/>

SamTov replied:
<p>The peopl at the summer school may not know what SMILES is. For the sake of understanding what "CCO" means, it is probably worth just putting a picture of ethanol in and showing how the SMILES string is built out of it.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4982/discussion/#comment-1730756654'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='8'/>

SamTov replied:
<p>Why not have a paragraph here about these computations? How they are done along with some plot of something. It would be good to understand the models more and to see what they can do. We can generate a small trajectory, for example, and have them plot something about it.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4982/discussion/#comment-1730756656'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

SamTov replied:
<p>Why not show it?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4982/discussion/#comment-1730756658'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

SamTov replied:
<p>Explain each line, especially the time units.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4982/discussion/#comment-1730756659'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='12'/>

SamTov replied:
<pre class="ql-syntax" spellcheck="false">Line #1.    for atom in etoh:
</pre><p>It isn't clear why the atoms should be in the model object. Have you introduced the type mapping? What it is, why you need it?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4982/discussion/#comment-1730756660'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='14'/>

SamTov replied:
<p>Should you tell people to open it in a different window as they will be running the descent after the opening? Maybe go to split screen or something?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4982/discussion/#comment-1730756662'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='15'/>

SamTov replied:
<ul><li>We haven't seen the first parts yet. Up to now, I don't know if it has accurate forces or can run MD. There is no comparison. You can do a classical force-field comparison, for example, or show DFT data. I actually have DFT ethanol trajectories you can use. Or, run a short DFT ethanol MD with PySCF, it is pretty fast.</li><li>Maybe list some chemical reaction methods.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4982/discussion/#comment-1730756664'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

SamTov replied:
<pre class="ql-syntax" spellcheck="false">Line #3.    system.thermostat.turn_off()
</pre><p>Why? People will ask why you are doing things like this. If you had to ask someone for help in making the script, the people in the tutorial will be at the same level but also without the ML knowledge. Explain everything.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4982/discussion/#comment-1730756665'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

SamTov replied:
<p>Write up a conlcusion, what have we learned, what can we do next, why is this cool.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4982/discussion/#comment-1730756667'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='38'/>

SamTov commented on doc/tutorials/mlip/mlip.ipynb:
```diff
@@ -0,0 +1,447 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    etoh = smiles2atoms("CCO")
</pre><p>I don't find etoh to be a clear variable name.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4982/discussion/#comment-1730756655'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='8'/>

PythonFZ replied:
<p><code>EtOH</code> is the most commonly used abbreviation for ethanol, but I renamed in to <code>ethanol</code></p>

SamTov commented:
Let us consider what we want from the tutorial. Is a reaction the only analysis we want to do? Don't most people want to run MD before doing a reaction study? Both are nice, but we could do RDF calculations or even Rg computations on something like ethanol. Or go for a bigger molecular and do a quick polymer study.

PythonFZ reviewed (commented):
Is there any reason you removed the interactive plotting part for the visualizer `vis.zndraw.figures = {"distance": fig.to_json()}`?

SamTov commented:
> Is there any reason you removed the interactive plotting part for the visualizer `vis.zndraw.figures = {"distance": fig.to_json()}`?

It didn't work when we ran it on the CIP pool computers.

PythonFZ commented:
> > Is there any reason you removed the interactive plotting part for the visualizer `vis.zndraw.figures = {"distance": fig.to_json()}`?
> 
> It didn't work when we ran it on the CIP pool computers.



> > Is there any reason you removed the interactive plotting part for the visualizer `vis.zndraw.figures = {"distance": fig.to_json()}`?
> 
> It didn't work when we ran it on the CIP pool computers.

I see - I opened https://github.com/espressomd/espresso/issues/4993 which would resolve this and make it work.

SamTov commented:
> > > Is there any reason you removed the interactive plotting part for the visualizer `vis.zndraw.figures = {"distance": fig.to_json()}`?
> > 
> > 
> > It didn't work when we ran it on the CIP pool computers.
> 
> > > Is there any reason you removed the interactive plotting part for the visualizer `vis.zndraw.figures = {"distance": fig.to_json()}`?
> > 
> > 
> > It didn't work when we ran it on the CIP pool computers.
> 
> I see - I opened #4993 which would resolve this and make it work.

Is it necessary? They can plot the plotly figure in the notebook and the syntax to have it opened in ZnDraw is quite odd, particularly for those who aren't familiar with the program.

PythonFZ commented:
> > > > Is there any reason you removed the interactive plotting part for the visualizer `vis.zndraw.figures = {"distance": fig.to_json()}`?
> > > 
> > > 
> > > It didn't work when we ran it on the CIP pool computers.
> > 
> > 
> > > > Is there any reason you removed the interactive plotting part for the visualizer `vis.zndraw.figures = {"distance": fig.to_json()}`?
> > > 
> > > 
> > > It didn't work when we ran it on the CIP pool computers.
> > 
> > 
> > I see - I opened #4993 which would resolve this and make it work.
> 
> Is it necessary? They can plot the plotly figure in the notebook and the syntax to have it opened in ZnDraw is quite odd, particularly for those who aren't familiar with the program.

- `opened in ZnDraw is quite odd` could you elaborate what is odd about it and ideally give a suggestion on how to improve it?
- I would like to showcase the interactive part of having plots alongside the data in the visualizer. It is not required but I thought it is a nice touch. From the MLIP standpoint it is not required.

SamTov reviewed (approved):
I like it.

PythonFZ commented:
Thanks for all the work you put into this 👍 

The
```
#TODO: figure out if this works
```
as well as the `vis.zndraw.figures = {"distance": fig.to_json()}` might both depend on https://github.com/espressomd/espresso/issues/4993

Unfortunately, I don't know what issues came up with this version and they did not mention or make any issues on the ZnDraw repo. If the issue will not be resolved before the tutorial it might be easiest to remove this part.

jngrad commented on doc/tutorials/mlip/mlip.ipynb:
```diff
@@ -0,0 +1,722 @@
+   "source": [
+    "**Exercise:**\n",
+    "- Take a look at the energies and the molecule in the visualizer during the minimisation. \n",
+    "- How does the energy compare to the literature value of ~3341.86 eV for the ground state?"
```
* this energy value needs a citation
* this energy value is not reproduced by the MACE-MP-0 model in the current form of the tutorial

SamTov replied:
We will remove this as we have no good explanation for why it does not come out right. It could be due to the temperature or a spurious comparison value.

SamTov replied:
Energy is now resolved and correct.

jngrad replied:
I'm still uneasy about users being asked to compare the ground state energy of ice against the ground state energy of water vapor. There is an 0.5 eV energy difference. One would need to draw the thermodynamic circle and add the enthalpy of vaporisation, enthalpy of fusion, and the integral of the specific heat capacity, which is more work than providing the reference ground state energy of water vapor from a paper.

jngrad commented on doc/tutorials/mlip/mlip.ipynb:
```diff
@@ -0,0 +1,722 @@
+    "**Exercise:**\n",
+    "- Indentify the highlighted atoms in the visualization window, which are the two hydrogen atoms of the sulfuric acid.\n",
+    "- Observe the moment they undergo the protonation event visually and also in the hydrogen-oxygen distance plots.\n",
+    "- Can you also see changes in the potential energy? Why is that?"
```
It's unclear to me, what is the answer to this question. The dip in energy could be due to the exothermic reaction $\mathrm{H_2SO_4 + H_2O \to HSO_4^{-} + H_3O^{+}}$, but the energy difference doesn't reproduce experimentally determined Gibbs free energy values. I'm also unsure if the potential energy can be directly equated to a Gibbs free energy, a Helmholtz free energy,  an enthalpy or an internal energy.

SamTov replied:
I would actually like to discuss this point during the tutorial. The question of whether this is a physically realistic process is fundamental to the explainability of ML potentials. During the discussion, I would also ask how to extend the study to determine the answer. Namely, running it longer and generating some statistics or pH measurements that can be compared to experimental values.

reinaual commented on doc/tutorials/mlip/mlip.ipynb:
```diff
@@ -0,0 +1,715 @@
+    "RDKit, the tool mentioned earlier, provides powerful utilities for creating 3D structures from these string representations of molecules.\n",
+    "\n",
+    "In the first part of the tutorial, we will perform a geometry optimization for water. \n",
+    "The chemical formula of ethanol is H<sub>2</sub>O."
```
```suggestion
    "The chemical formula of water is H<sub>2</sub>O."
```
It's not ethanol anymore

PythonFZ replied:
fixed everywhere.

RudolfWeeber commented:
Is something till missing or can ew go ahed?

SamTov commented:
The ground state energies for the water study are resolved and correct.

PythonFZ committed:
MLIP tutorial

Co-authored-by: Alexander Reinauer <areinauer@icp.uni-stuttgart.de>
Co-authored-by: Samuel Tovey <tovey.samuel@gmail.com>
Co-authored-by: Jean-Noël Grad <jgrad@icp.uni-stuttgart.de>
